### PR TITLE
add WString support

### PR DESCRIPTION
**Pull request by dirk-thomas**
Connect to #352.

While waiting for ros2/rcl_interfaces#58 the patch (without tests) can already be reviewed.

### CODE REVIEW

File: rosidl_generator_c/src/u16string_functions.c [hunk 1]
```diff
@@ -115,6 +127,24 @@ rosidl_generator_c__U16String__len(const uint16_t * value)
...
 }
 
+bool
+rosidl_generator_c__U16String__resize(
+  rosidl_generator_c__U16String * str, size_t n)
+{
+  if (!str) {
+    return false;
+  }
+  uint16_t * data = realloc(str->data, (n + 1) * sizeof(uint16_t));
```
Comment (jacobperron) [inline rosidl_generator_c/src/u16string_functions.c]: Similar to the string implementation, consider guarding against `n == SIZE_MAX`:

https://github.com/ros2/rosidl/blob/4dce828f1a0b021d3e16b97e560259a9201be862/rosidl_generator_c/src/string_functions.c#L81-L84

Comment (dirk-thomas) [reply rosidl_generator_c/src/u16string_functions.c]: Good point: see d9e6fad36892174ad0b1940053dc7641acde5f45.

File: rosidl_generator_c/test/test_interfaces.c [hunk 2]
```diff
@@ -333,8 +337,39 @@ int test_strings(void)
...
+{
+  bool res = false;
+  rosidl_generator_c__msg__WStrings * wstrings = NULL;
+
+  wstrings = rosidl_generator_c__msg__WStrings__create();
+  EXPECT_NE(wstrings, NULL);
+
+  res = rosidl_generator_c__U16String__assign(&wstrings->empty_wstring, TEST_WSTRING);
+  EXPECT_EQ(true, res);
+  // EXPECT_EQ(0, strcmp(wstrings->empty_wstring.data, TEST_WSTRING));
```
Comment (jacobperron) [inline rosidl_generator_c/test/test_interfaces.c]: What are the issues with these tests? Maybe add a TODO explaining.

File: rosidl_generator_c/test/test_interfaces.c [hunk 3]
```diff
@@ -333,8 +337,39 @@ int test_strings(void)
...
+  // EXPECT_EQ(0, strcmp(wstrings->def_wstring.data, "Hello world!"));
+  // EXPECT_EQ(0, strcmp(wstrings->def_wstring2.data, "Hello'world!"));
+  // EXPECT_EQ(0, strcmp(wstrings->def_wstring3.data, "Hello\"world!"));
+  // EXPECT_EQ(0, strcmp(wstrings->def_wstring4.data, "Hello'world!"));
+  // EXPECT_EQ(0, strcmp(wstrings->def_wstring5.data, "Hello\"world!"));
+  // since upper-bound checking is not implemented yet, we restrict the string copying
+  // TODO(mikaelarguedas) Test string length properly instead of cheating copy
```
Comment (jacobperron) [inline rosidl_generator_c/test/test_interfaces.c]: dirk-thomas

File: rosidl_generator_c/test/test_interfaces.c (hunk 2 above)
Comment (dirk-thomas) [reply rosidl_generator_c/test/test_interfaces.c]: Added in 45b923db0afc7f164473b84f9c61f4c0ad94dc6c.

That will also cover this comment: https://github.com/ros2/test_interface_files/pull/4#discussion_r280528815

File: rosidl_generator_c/test/test_interfaces.c (hunk 3 above)
Comment (dirk-thomas) [reply rosidl_generator_c/test/test_interfaces.c]: 1-to-1 copy from the `String` tests above. So I will keep it as is.

**Pull request merged**